### ISSUE TITLE
ArgumentError (not fully converted, 23 bytes left) with String.encode!

### ISSUE THREAD

**Issue by bibstha** [jruby/jruby#5681]
### Environment

Provide at least:
* JRuby version (`jruby -v`) and command line (flags, JRUBY_OPTS, etc)
`jruby 9.2.6.0 (2.5.3) 2019-02-11 15ba00b OpenJDK 64-Bit Server VM 12+33 on 12+33 +jit [darwin-x86_64]`

* Operating system and platform (e.g. `uname -a`)
MacBookPro Mojave 

### Problem:

I'm trying to read and convert the following file in utf8
https://github.com/tmm1/ripper-tags/blob/7f31ab7d9009ea2c566e81901cd344b04e6356e1/test/fixtures/encoding.rb

Here is the ruby code that works fine in cruby

```
# test_encoding.rb
filename = "test/fixtures/encoding.rb"
str = File.open(filename, "r:utf8") { |f| f.read }
str.encode!('utf-16', :invalid => :replace, :undef => :replace)
```

On jruby it raises an exception

```
$ ruby test_encoding.rb
ArgumentError: not fully converted, 23 bytes left
  encode! at org/jruby/RubyString.java:6044
   <main> at test.rb:3
```

**Comment by headius** [comment]
Strange...I am also on a MacBook and I cannot get this to fail. I wget the encoding.rb file and run your script against it, and only get a warning:

```
$ jruby blah.rb
blah.rb:3: warning: Unsupported encoding utf8 ignored
```

Fixing that to be `utf-8` makes it complete without any warnings or errors.

What is the output of this script for you?

```ruby
puts ENV_JAVA['file.encoding']
```

**Comment by kares** [comment]
works on Linux x86_64 as well, using Java 8/11 also tried *OpenJDK 12.0.1+12* (same built as yours)

**Comment by headius** [comment]
@bibstha We will need to get some information from you about your environment, since we are unable to reproduce locally. Mostly we need to know if there's a different system encoding, so we can sort out which encoding is not working right.

**Comment by bibstha** [comment]
Weird I can reproduce this on archlinux too now.

```bash
#! /bin/bash
wget https://raw.githubusercontent.com/tmm1/ripper-tags/7f31ab7d9009ea2c566e81901cd344b04e6356e1/test/fixtures/encoding.rb

tee bug.rb <<EOF
filename = "encoding.rb"
str = File.open(filename, "r:utf-8") { |f| f.read }
str.encode!('utf-16', :invalid => :replace, :undef => :replace)
EOF

jruby bug.rb
```

```
$ jruby --version
jruby 9.2.7.0 (2.5.3) 2019-04-09 8a269e3 OpenJDK 64-Bit Server VM 11.0.3+4 on 11.0.3+4 +jit [linux-x86_64]
```

When I add `puts ENV_JAVA['file.encoding']`, it says `ANSI_X3.4-1968` on archlinux.
On osX Mojave, `ENV_JAVA['file.encoding']` says `UTF-8`

```
$ jruby bug.rb
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.jruby.util.SecurityHelper to field java.lang.reflect.Field.modifiers
WARNING: Please consider reporting this to the maintainers of org.jruby.util.SecurityHelper
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release

ArgumentError: not fully converted, 23 bytes left
  encode! at org/jruby/RubyString.java:6048
   <main> at bug.rb:4
```

**Comment by headius** [comment]
@bibstha Yeah I ran your bash script and it still passes for me. I tried to force the encoding to that weird ANSI string but could not get it to fail.

```
[] ~/projects/jruby $ sh bug.bash 
--2019-04-26 12:06:13--  https://raw.githubusercontent.com/tmm1/ripper-tags/7f31ab7d9009ea2c566e81901cd344b04e6356e1/test/fixtures/encoding.rb
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.184.133
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.184.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 77 [text/plain]
Saving to: ‘encoding.rb.4’

encoding.rb.4                                       100%[=================================================================================================================>]      77  --.-KB/s    in 0s      

2019-04-26 12:06:13 (2.94 MB/s) - ‘encoding.rb.4’ saved [77/77]

filename = "encoding.rb"
str = File.open(filename, "r:utf-8") { |f| f.read }
str.encode!('utf-16', :invalid => :replace, :undef => :replace)

[] ~/projects/jruby $ LANG=ANSI_X3.4-1968 sh bug.bash 
--2019-04-26 12:06:19--  https://raw.githubusercontent.com/tmm1/ripper-tags/7f31ab7d9009ea2c566e81901cd344b04e6356e1/test/fixtures/encoding.rb
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.184.133
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.184.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 77 [text/plain]
Saving to: 'encoding.rb.5'

encoding.rb.5                                       100%[=================================================================================================================>]      77  --.-KB/s    in 0s      

2019-04-26 12:06:19 (2.16 MB/s) - 'encoding.rb.5' saved [77/77]

filename = "encoding.rb"
str = File.open(filename, "r:utf-8") { |f| f.read }
str.encode!('utf-16', :invalid => :replace, :undef => :replace)
```

Can you try to reproduce this in a Docker container or something?

Also meanwhile, you could pass `-Xbacktrace.style=full` to JRuby, so we can see the exact line where it's failing.

**Comment by k77ch7** [comment]
@headius  I created a [Dockerfile](https://gist.githubusercontent.com/k77ch7/006ec9c07ff82ecab94d6f61de495fb6/raw/c3b013634b8da4e6e334403c7acf864294fcaa5a/Dockerfile) to reproduce this.
```
~/tmp $ curl -O https://gist.githubusercontent.com/k77ch7/006ec9c07ff82ecab94d6f61de495fb6/raw/c3b013634b8da4e6e334403c7acf864294fcaa5a/Dockerfile
~/tmp $ docker build .
```

**Comment by headius** [comment]
Reproduced!

**Comment by headius** [comment]
I may see a possible bug. It seems like the code may not be incrementing a byte count when inserting the replacement characters along some encoding paths.

**Comment by headius** [comment]
Ok so there's at least a workaround.

The code you have is reading in the file as UTF-8 (using the `"r:utf-8"` flag to `File.read`) and so the resulting content – which is NOT valid UTF-8 – ends up looking like a UTF-8 string with bad UTF-8 content in it. On my local system, this seems to handle the undef/invalid replacement ok and produces a string with some `?` in it. On your docker config, this does not seem to happen; the transcoding stops after the first bad byte, resulting in your error.

The workaround is to force the string's encoding to be what you expect it to be (e.g. `str.force_encoding("ISO-8859-1")`, and then it transcodes correctly. Of course you may not be able to tell that encoding when reading the files, or you may want the replacement logic to fire and just assume everything is UTF-8.

The bug here seems to be that we bail out of the transcoding and NOT do the replacements, but only under some environments.

I'm pretty sure this is some sort of mismatch between the system encoding, the JDK encoding, and our Ruby encoding logic, but I have not been able to solve it (or reproduce it) using the environment variables and properties I'd expect to solve it.

**Comment by headius** [comment]
I think the next step for debugging will have to be expanding this docker setup to allow downloading and building JRuby master, along with perhaps hooking up JVM debugging. I'm stumped at the moment.

**Comment by headius** [comment]
Well something seems to have fixed this in JRuby 9.3, probably the recent fixes to jcodings by @k77ch7 and others. Hooray!

**Comment by headius** [comment]
Confirmed working by shelling into the docker container and using JRuby master to test it.
